src/main/java/com/github/onsdigital/JWTKeyProviderImpl.java

```diff
@@ -0,0 +1,94 @@
+     *
+     * @param url             used to fetch the signing keys
+     * @param initialInterval the initial interval in milliseconds to be used for exponential retries
+     * @param maxElapsedTime  the max elapsed time in milliseconds to be used for exponential retries
+     * @param maxInterval     the max interval in milliseconds to be used for exponential retries
+     * @throws IllegalArgumentException if the public signing keys provided are invalid
+     */
+    JWTKeyProviderImpl(String url, int initialInterval, int maxElapsedTime, int maxInterval) {
+        this.identityApiUrl = url;
+        this.initialInterval = initialInterval;
+        this.maxElapsedTime = maxElapsedTime;
+        this.maxInterval = maxInterval;
+    }
+
+
+    /**
+     * Fetches singing keys from identity api server, returning valid signing keys on success api response.
+     *
+     * @return the singing keys used for decoding JWT tokens
+     * @throws Exception if failed to fetch signing keys from the server
+     */
+    @Override
+    public Map<String, String> getJwtKeys() throws Exception {
+        HttpResponse response = null;
+        try {
+            response = getJwtKeysFromIdentityApi();
+
+            if (response == null) {
+                throw new Exception("Failed to get response from server:" + identityApiUrl);
+            }
+
+            if (response.getStatusCode() != HttpStatusCodes.STATUS_CODE_OK) {
+                throw new Exception("Failed to get jwt keys:" + response.parseAsString());
+            }
+            Map<String, String> jwtKeys = getJwtKeysFromResponse(response);
+            if (jwtKeys.isEmpty()) {
+                throw new Exception("JWT keys not found in the response");
+            }
+            return jwtKeys;
+        } finally {
+            if (response != null) {
+                response.disconnect();
+            }
+        }
+    }
+
+    HttpResponse getJwtKeysFromIdentityApi() throws IOException {
+        ExponentialBackOff backoff = new ExponentialBackOff.Builder()
+                .setInitialIntervalMillis(initialInterval)
+                .setMaxElapsedTimeMillis(maxElapsedTime)
+                .setMaxIntervalMillis(maxInterval)
+                .build();
+
+        HttpRequest httpRequest = new NetHttpTransport()
+                .createRequestFactory()
+                .buildGetRequest(new GenericUrl(identityApiUrl))
+                .setUnsuccessfulResponseHandler(new HttpBackOffUnsuccessfulResponseHandler(backoff));
+        return httpRequest.execute();
+    }
+
+    private Map<String, String> getJwtKeysFromResponse(HttpResponse response) throws IOException {
+        Map<String, String> signingKeys = new HashMap<>();
+        response.parseAs(Data.newMapInstance(HashMap.class).getClass())
+                .forEach((key, value) -> signingKeys.put(String.valueOf(key), String.valueOf(value)));
+        return signingKeys;
+    }
+
+}
```

Comment:
Now when we test this class now we can mock `RequestBuilder` and can control how it behaves allowing us to test all the different scenarios.